File: python-alert-service/models/alert_rule.py
"""Pydantic models for alert rule data."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class Operator(str, Enum):
    """Valid comparison operators for alert thresholds."""

    GT = "gt"
    LT = "lt"
    GTE = "gte"
    LTE = "lte"
    EQ = "eq"


class RuleStatus(str, Enum):
    """Valid alert rule statuses."""

    ACTIVE = "active"
    INACTIVE = "inactive"


class AlertRuleBase(BaseModel):
    """Base alert rule model with common fields."""

    sensor_id: str = Field(..., min_length=1, description="ID of the sensor to monitor")
    metric: str = Field(default="value", min_length=1, max_length=50, description="Metric to evaluate")
    operator: Operator = Field(..., description="Comparison operator")
    threshold: float = Field(..., description="Threshold value")
    name: str = Field(..., min_length=1, max_length=200, description="Rule display name")
    status: RuleStatus = Field(default=RuleStatus.ACTIVE, description="Rule status")


class AlertRuleCreate(AlertRuleBase):
    """Model for creating a new alert rule."""

    pass


class AlertRuleUpdate(BaseModel):
    """Model for updating an existing alert rule. All fields optional."""

    sensor_id: Optional[str] = Field(None, min_length=1)
    metric: Optional[str] = Field(None, min_length=1, max_length=50)
    operator: Optional[Operator] = None
    threshold: Optional[float] = None
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    status: Optional[RuleStatus] = None


class AlertRule(AlertRuleBase):
    """Complete alert rule model with ID and timestamps.

    Both timestamps are always populated: updated_at is set to
    created_at on initial creation, matching the sensor service pattern.
    """

    id: str = Field(..., description="Unique rule identifier")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp (equals created_at on creation)")

    class Config:
        from_attributes = True


class AlertRuleResponse(BaseModel):
    """Response model for alert rule creation (may include warning)."""

    id: str
    sensor_id: str
    metric: str
    operator: Operator
    threshold: float
    name: str
    status: RuleStatus
    created_at: str
    updated_at: str
    warning: Optional[str] = None

    class Config:
        from_attributes = True


class AlertRuleList(BaseModel):
    """Response model for list of alert rules."""

    rules: list[AlertRule]
    count: int

File: go-alert-service/models/alert_rule.go
package models

import (
	"fmt"
	"time"
)

// ValidOperators contains all allowed comparison operators.
var ValidOperators = map[string]bool{
	"gt":  true,
	"lt":  true,
	"gte": true,
	"lte": true,
	"eq":  true,
}

// ValidRuleStatuses contains all allowed rule statuses.
var ValidRuleStatuses = map[string]bool{
	"active":   true,
	"inactive": true,
}

// AlertRule represents a complete alert rule record from the database.
type AlertRule struct {
	ID        string  `json:"id"`
	SensorID  string  `json:"sensor_id"`
	Metric    string  `json:"metric"`
	Operator  string  `json:"operator"`
	Threshold float64 `json:"threshold"`
	Name      string  `json:"name"`
	Status    string  `json:"status"`
	CreatedAt string  `json:"created_at"`
	UpdatedAt string  `json:"updated_at"`
}

// AlertRuleCreate represents the request body for creating a new alert rule.
type AlertRuleCreate struct {
	SensorID  string  `json:"sensor_id" binding:"required,min=1"`
	Metric    string  `json:"metric"`
	Operator  string  `json:"operator" binding:"required"`
	Threshold float64 `json:"threshold"`
	Name      string  `json:"name" binding:"required,min=1,max=200"`
	Status    string  `json:"status"`
}

// Validate checks if the AlertRuleCreate fields are valid.
func (r *AlertRuleCreate) Validate() error {
	if !ValidOperators[r.Operator] {
		return fmt.Errorf("invalid operator: %s", r.Operator)
	}
	if r.Status == "" {
		r.Status = "active"
	}
	if !ValidRuleStatuses[r.Status] {
		return fmt.Errorf("invalid rule status: %s", r.Status)
	}
	if r.Metric == "" {
		r.Metric = "value"
	}
	return nil
}

// AlertRuleUpdate represents the request body for updating an alert rule.
// All fields are optional (pointers).
type AlertRuleUpdate struct {
	SensorID  *string  `json:"sensor_id,omitempty"`
	Metric    *string  `json:"metric,omitempty"`
	Operator  *string  `json:"operator,omitempty"`
	Threshold *float64 `json:"threshold,omitempty"`
	Name      *string  `json:"name,omitempty"`
	Status    *string  `json:"status,omitempty"`
}

// Validate checks if the AlertRuleUpdate fields are valid.
func (r *AlertRuleUpdate) Validate() error {
	if r.Operator != nil && !ValidOperators[*r.Operator] {
		return fmt.Errorf("invalid operator: %s", *r.Operator)
	}
	if r.Status != nil && !ValidRuleStatuses[*r.Status] {
		return fmt.Errorf("invalid rule status: %s", *r.Status)
	}
	return nil
}

// AlertRuleResponse extends AlertRule with an optional warning field.
// The warning is populated when the sensor service is unavailable during
// rule creation (circuit breaker fallback).
type AlertRuleResponse struct {
	ID        string  `json:"id"`
	SensorID  string  `json:"sensor_id"`
	Metric    string  `json:"metric"`
	Operator  string  `json:"operator"`
	Threshold float64 `json:"threshold"`
	Name      string  `json:"name"`
	Status    string  `json:"status"`
	CreatedAt string  `json:"created_at"`
	UpdatedAt string  `json:"updated_at"`
	Warning   string  `json:"warning,omitempty"`
}

// AlertRuleList represents the response for listing alert rules.
type AlertRuleList struct {
	Rules []AlertRule `json:"rules"`
	Count int         `json:"count"`
}

// Now returns the current UTC time as an ISO 8601 string.
func Now() string {
	return time.Now().UTC().Format(time.RFC3339)
}
